### ISSUE TITLE
Bactopia v2 Overview

### ISSUE THREAD

**Issue by rpetit3** [bactopia/bactopia#233]
# Bactopia v2 Overview
With tremendous effort by @mxrcon and @abhi18av, the [foundation for migrating Bactopia to DSL2 has been laid out](https://github.com/bactopia/bactopia/pull/228). This transition represents the key milestone to push Bactopia to v2! (Super excited about this Davi and Abhinav!)

By switching to DSL2, the door for creating custom Bactopia workflows has been opened. For example, let's say you have some *Staphylococcus aureus* samples, and you want to run Bactopia and then the Bactopia Tool [staph-typer](https://bactopia.github.io/bactopia-tools/staph-typer/). Instead with DSL2, we can create a sub-workflow (e.g. _[Staphopia](https://staphopia.emory.edu/)_) that will automatically run Bactopia and staph-typer. In other words, we can start creating organism-specific sub-workflows, as well as sub-workflows that only include certain steps such as assembly.

I think this also a good time to start cleaning up some things and adding features that will make long-term maintenance more sustainable.

### House Cleaning
These are to help reduce the burden required to maintain Bactopia long-term. These are really about standardizing things in such a way that we can automate things. For example, printing usage across each of the workflows can be configured through config files (e.g. [nf-core json schema](https://github.com/nf-core/sarek/blob/master/nextflow_schema.json). There are also a lot of shared functions for checking inputs, creating channels, etc. These duplications are no longer necessary in DSL2.

- Automate Version and Citation tracking
    - [x] Add meta.yml to every module 
        - Use [nf-core/modules meta.yml](https://github.com/nf-core/modules/blob/master/modules/artic/minion/meta.yml) a template
    - [x] Integrate this into `bactopia citations`
    - [x] Integrate this into `bactopia versions` (now handled by `versions.yml`)

- Reduce code duplication
    - [x] Schema for printing Usage (_using modified nf-core pipelines `lib/*`_)
    - [x] Input file checking (_using modified nf-core pipelines `lib/*`_)
    - [x] Merging tables https://github.com/nf-core/modules/pull/785
    - [x] Input validation (e.g. type checking)  (_using modified nf-core pipelines lib/*_)

- Organize DSL2 structure
    - [x] Create structure to follow (_using structure loosely based off nf-core dsl2 pipelines_)
    
- Fixes
    - [x] https://github.com/bactopia/bactopia/issues/212
    - [x] https://github.com/bactopia/bactopia/issues/240
    - [x] https://github.com/bactopia/bactopia/issues/245

### Additional Features

- [x] Support for Nanopore reads
- [X] Drop support for some uncompressed outputs (e.g. assemblies)
    - Defaults to compressed outputs, `--skip_compression` disables this feature
- [x] GenBank compatible assembly
    - Currently does not like `gnl|`  
- [x] Outputs from the tutorial https://doi.org/10.6084/m9.figshare.17097156.v1

### Implement `pytest` for testing
I'd like to create a suite of tests that are operated by `pytest` and `pytest-workflows`. The `nf-core/modules` team has a [framework](https://github.com/nf-core/modules#running-tests-manually) that can be extended to Bactopia.

- [x] Setup Test-Data repo - _Done! https://github.com/bactopia/bactopia-tests_
- [x] Setup walk through for testing - _Done! https://github.com/bactopia/bactopia/tree/dsl2/tests_
- [x] Add tests to Github Actions -_Done! https://github.com/bactopia/bactopia/actions/runs/1256507737_
- Create Tests for Bactopia Modules
    - [x] `annotate_genome`
    - [x] `antimicrobial_resistance`
    - [x] `ariba_analysis`
    - [x] `assemble_genome`
    - [x] `assembly_qc`
    - [x] `blast`
    - [x] `call_variants`
    - [x] ~~`count_31mers`~~ (_merged into `minmer_sketch`_)
    - [x] ~~`download_references`~~ (_merged into `call_variants`_)
    - [x] ~~`estimate_genome_size`~~ (_merged into `gather_samples`_)
    - [x] ~~`fastq_status`~~ (_merged into `gather_samples`_)
    - [x] `gather_samples`
    - [x] `mapping_query`
    - [x] `minmer_query`
    - [x] `minmer_sketch`
    - [x] `qc_reads`
    - [x] `sequence_type`
- Create Tests for Bactopia Tool Modules
    - [x] `agrvate` 
    - [x] `bakta`
    - [x] `ectyper`
    - [x] `emmtyper`
    - [x] `eggnog`
    - [x] `fastani`
    - [x] `hicap`
    - [x] `ismapper`
    - [x] `kleborate`
    - [x] `lissero`
    - [x] `mashtree`
    - [x] `meningotype`
    - [x] `ngmaster`
    - [x] `pangenome`
    - [x] `seqsero2`
    - [x] `spatyper`
    - [x] `staph-typer`
    - [x] `staphopiasccmec`
    - [x] `tbprofiler`
  
### Convert some processes to `nf-core/modules`
There are a few tools used by Bactopia that are the only tool in the process. Most of these tools are in the Bactopia Tools. I think its best that these tools be transferred to `nf-core/modules`. Many of these will need to be added to `nf-core` but they are in need of some bacterial genomic tool love, so its ok!

- Need to be added (25 total)
    - [x] `agrvate` https://github.com/nf-core/modules/pull/693
    - [x] `bakta` https://github.com/nf-core/modules/pull/1085
    - [x] `clonalframe` https://github.com/nf-core/modules/pull/974
    - [x] `ectyper` https://github.com/nf-core/modules/pull/948
    - [x] `eggnog_mapper` https://github.com/nf-core/modules/pull/1020
    - [x] `emmtyper` https://github.com/nf-core/modules/pull/1028
    - [x] `fastani` https://github.com/nf-core/modules/pull/695
    - [x] `fastq-scan` https://github.com/nf-core/modules/pull/935
    - [x] `gtdb` https://github.com/nf-core/modules/pull/765
    - [x] `hicap` https://github.com/nf-core/modules/pull/772
    - [x] `ismapper` https://github.com/nf-core/modules/pull/773
    - [x] `kleborate` https://github.com/nf-core/modules/pull/711
    - [x] `lissero` https://github.com/nf-core/modules/pull/1026
    - [x] `mashtree` https://github.com/nf-core/modules/pull/767
    - [x] `meningotype` https://github.com/nf-core/modules/pull/1022
    - [x] `ngmaster` https://github.com/nf-core/modules/pull/1024
    - [x] `phyloflash` https://github.com/nf-core/modules/pull/786
    - [x] `pirate` https://github.com/nf-core/modules/pull/777
    - [x] `roary` https://github.com/nf-core/modules/pull/776
    - [x] `scoary` https://github.com/nf-core/modules/pull/1034
    - [x] `seqsero2` https://github.com/nf-core/modules/pull/1016
    - [x] `snp-dists` https://github.com/nf-core/modules/pull/697
    - [x] `spatyper` https://github.com/nf-core/modules/pull/784
    - [x] `staphopia-sccmec` https://github.com/nf-core/modules/pull/702
    - [x] `tb-profiler` https://github.com/nf-core/modules/pull/947

### Curated Datasets
I think one of the best features of Bactopia is the ability to include public datasets. This works great for general datasets, but organism-specific datasets are kind of lost. I think it would be great to start a set of curated datasets that users can add data to.

Here's an example of a curated [Staphylococcus aureus Bactopia Dataset](https://github.com/bactopia-datasets/staphylococcus-aureus). This dataset can easily be imported and allow users to rapidly analyze their samples with a curated dataset specific to their organism.

I think it would also be nice if these curated datasets, included SRA accessions linked to publications. But this exceeds my capabilities and would require extensive community support.

### Species specific Workflows
With DSL2, we can create _Species Specific_ workflows by combining the main Bactopia workflow with some Bactopia Tools. The main example, and thus shall act as a _proof-of-concept_ will be Staphopia. Staphopia is essentially Bactopia + the Bactopia Tool `staph-typer`.

- [x] Create a _Staphopia_ workflow

**Comment by rpetit3** [comment]
This is a Work-In-Progress, and evolving. Please feel free to provide any feedback and requests.

**Comment by Mxrcon** [comment]
@rpetit3 I completely agree with the design pattern that you're aiming to implement. Creating specific workflows for some genus makes complete sense and could be easily implemented with DSL2. Let me know if i can make myself useful on those tasks!

**Comment by abhi18av** [comment]
Regarding the inclusion of outlined modules into nf-core, @Mxrcon  and I are happy to take this forward as discussed on slack.

I'm dividing these into two groups, for initiating a baseline draft module (we'll be collaborating extensively of course) 

@abhi18av

- agrvate
- clonalframe
- ectyper
- eggnog_mapper
- fastani
- gtdb
- hicap
- ismapper


@Mxrcon 

- kleborate
- mashtree
- phyloflash
- pirate
- roary
- seqsero2
- snp-dists
- spatyper
- staphopia-sccmec

P.S. @rpetit3 , if possible could you please add us as collaborators, so that we could start making use of https://github.com/bactopia/bactopia/projects/1 as well?

**Comment by kusandeep** [comment]
Looking forward to Bactopia v2.
I am already excited to see those already listed tool in v2. Additionally, I am interested in   #157 ,  pangenome visualisation tools would be great too #156 and Scoary (https://github.com/AdmiralenOla/Scoary) #159

**Comment by abhi18av** [comment]
@rpetit3 , I need a bit of guidance here regarding the contributions to nf-core modules.


1. Which sub-commands exactly do we need from the tools mentioned here https://github.com/bactopia/bactopia/issues/233#issuecomment-899593455 . This is necessary to understand since the structure of modules is `TOOL/SUB_COMMAND`
2. The `nf-core/modules` generally rely on https://github.com/nf-core/test-datasets/ repo,  which only has `human` and `sars-cov` datasets, we might need to decide which bacterial dataset we want to add to test these modules. I'm thinking `E. Coli` by default, but happy to hear your thoughts.

**Comment by rpetit3** [comment]
_1. Which sub-commands exactly do we need from the tools mentioned here Bactopia v2 Overview #233 (comment) . This is necessary to understand since the structure of modules is TOOL/SUB_COMMAND_

I'll get this put together, but I think most of them don't have subcommands

_2. The nf-core/modules generally rely on https://github.com/nf-core/test-datasets/ repo, which only has human and sars-cov datasets, we might need to decide which bacterial dataset we want to add to test these modules. I'm thinking E. Coli by default, but happy to hear your thoughts._

I noticed that as well. To get around this I created https://github.com/bactopia/bactopia-tests which is modeled after nf-core's test-datasets, just customized for bactopia. For these tests I'm using Candidatus Portiera aleyrodidarum which only has a genome size of 350kb. Here's some more info https://github.com/bactopia/bactopia/tree/dsl2/tests I've also modeled the [tests.config](https://github.com/bactopia/bactopia/blob/dsl2/conf/tests.config) after nf-core/modules (glad I was there for that hackathon!).

The small size genome allows me to test functionality quickly, but I think more importantly with limited resources (my desktop, and eventually GitHub Actions).

Now, I think eventually, another set of genomes (multiple organisms, E coli being one) will be needed as a validation set. For these it will be to make Bactopia is still given the same results, but also larger genomes are not producing errors we wouldn't have seen using a small genome. I think this would be a pre-release test set, but again something we could ideally get to fit on GitHub Action (not sure that'll be possible, unless the memory is increased >7gb on the Linux instance)

**Comment by rpetit3** [comment]
Small update. I've been doing some cleanup and  integrating more nf-core inspirations into v2. The clean up is to make things easier to maintain, and nf-core is because I think they are great practices to follow and by including them now would make a potential nf-core transition for Bactopia easier.

### `save_files` function for `publishDir`
I added a [`save_files`](https://github.com/bactopia/bactopia/blob/dsl2/modules/local/utilities/functions.nf#L39-L78) function that is inspired by [nf-core's saveFiles](https://github.com/nf-core/modules/blob/master/modules/abacas/functions.nf#L39-L68).   I opted to change the name slightly to denote the difference. `save_files` will move files to the appropriate output locations. It allows:

- you to ignore certain files
- rename the logs subdirectory (e.g. ariba for multiple datasets, `logs/ariba_analysis/card` `logs/ariba_analysis/vfdb`)
- places any `*-error.txt` files in the sample base directory
- uses [outputs.config](https://github.com/bactopia/bactopia/blob/dsl2/conf/outputs.config) to place outputs into directory by `PROCESS_NAME`
- renames `results/` directory to the `PROCESS_NAME` output
    - `results/` is a generic name for process outputs that go to a directory
    - This is also adopted from some nf-core workflows


Here's an example of how this looks for `qc_reads`
```
    publishDir "${params.outdir}/${sample}",
        mode: params.publish_mode,
        overwrite: params.overwrite,
        saveAs: { filename -> save_files(filename:filename, process_name:PROCESS_NAME, ignore:[ '-genome-size.txt', extra]) }

    output:
    tuple val(sample), val(single_end), path("results/${sample}*.fastq.gz"), emit: fastq, optional: true
    tuple val(sample), val(sample_type), val(single_end), path("results/${sample}*.fastq.gz"), path(extra), path(genome_size), emit: fastq_assembly, optional: true
    path "results/*"
    path "*.std{out,err}.txt", emit: logs, optional: true
    path ".command.*", emit: nf_logs
    path "*.version.txt", emit: version
    path "*-error.txt", optional: true
```

### Main Bactopia Process Cleanup
I've cleaned up the main Bactopia processes significantly.

- removed `check_staging` from all processes
- removed bash traps
- Nextflow's files (`.command.*`) are captured as outputs
- program versions are captured following [nf-core's modules](https://github.com/nf-core/modules/blob/master/modules/abacas/main.nf#L43) setup (e.g. `<SOFTWARE_NAME>.version.txt`)
- Where possible I've removed duplicate code which was most common on `single-end` vs `paired-end`
    - This is now handled by Nextflow variables, see [assemble_genome](https://github.com/bactopia/bactopia/blob/dsl2/modules/local/bactopia/assemble_genome/main.nf#L48-L85) or [qc_reads](https://github.com/bactopia/bactopia/blob/dsl2/modules/local/bactopia/qc_reads/main.nf#L39-L42)
- Capture program versions to `<SOFTWARE>.version.txt`
- Capture STDOUT and STDERR to `<SOFTWARE>.std{err,out}.txt`
- Updated all tests to check for the new outputs

### Process consolidation
While cleaning up, I consolidated a number of processes. In all, 11 processes were merged into other processes, taking the number of processes to maintain from 23 to 13. This was mostly done because these processes tended to be super quick (e.g. making a blast database) so this should reduce overall runtimes due to less overhead associated with starting and stopping processes. I think it also just makes things a bit easier to maintain.

- `blast_genes`, `blast_primers` and `blast_proteins` are now apart of `blast`
- `qc_original_summary` and `qc_final_summary` are now apart of `qc_reads`
- `download_references` and `call_variants_auto` are now apart of `call_variants`
- `fastq_status` and `estimate_genome_size` are now apart of `gather_samples`
- `make_blastdb` is now apart of `assemble_genome`
- `count_31mers` is now apart of `minmer_sketch`
    - Counting 31mers is optional and activated by `--count_31mers`

So the main workflow now looks like this:
![Bactopia Workflows (1)](https://user-images.githubusercontent.com/5334269/131876230-d8d6a1b8-7d87-4c8e-baa3-3a3e4d9693dc.png)

At this point, I think I'm all set to start making subworkflows.

**Comment by rpetit3** [comment]
Small update, I took the [`lib` folder from nf-core's tools pipeline template](https://github.com/nf-core/tools/tree/master/nf_core/pipeline-template/lib), and hacked it to work for Bactopia.

In doing so, this provides:
- arg parsing and validation
- usage printing
- framework for bactopia tools
- sharable schemas between workflows

It also keeps Bactopia on track for maintaining convergent evolution with nf-core practices, which would make any potential transitions easier.

```
nextflow run main.nf --help
N E X T F L O W  ~  version 21.04.0
Launching `main.nf` [marvelous_mirzakhani] - revision: 555ddf8329


------------------------------------------------------
   _                _              _
  | |__   __ _  ___| |_ ___  _ __ (_) __ _
  | '_ \ / _` |/ __| __/ _ \| '_ \| |/ _` |
  | |_) | (_| | (__| || (_) | |_) | | (_| |
  |_.__/ \__,_|\___|\__\___/| .__/|_|\__,_|
                            |_|
  bactopia v1.7.1
------------------------------------------------------
Typical pipeline command:

  bactopia --fastqs samples.txt --datasets datasets/ --species 'Staphylococcus aureus' -profile singularity

Required Parameters
  ### For Procesessing Multiple Samples
  --fastqs                     [string]  A FOFN with sample names and paths to FASTQ/FASTAs to process

  ### For Processing A Single Sample
  --R1                         [string]  First set of compressed (gzip) paired-end FASTQ reads (requires --R2 and --sample)
  --R2                         [string]  Second set of compressed (gzip) paired-end FASTQ reads (requires --R1 and --sample)
  --SE                         [string]  Compressed (gzip) single-end FASTQ reads  (requires --sample)
  --hybrid                     [boolean] Treat `--SE` as long reads for hybrid assembly.  (requires --R1, --R2, --SE and --sample)
  --sample                     [string]  Sample name to use for the input sequences

  ### For Downloading from SRA/ENA or NCBI Assembly
  **Note: Downloaded assemblies will have error free Illumina reads simulated for processing.**
  --accessions                 [string]  A file containing ENA/SRA Experiment accessions or NCBI Assembly accessions to processed
  --accession                  [string]  Sample name to use for the input sequences

  ### For Processing an Assembly
  **Note: Assemblies will have error free Illumina reads simulated for processing.**
  --assembly                   [string]  A assembled genome in compressed FASTA format. (requires --sample)

Dataset Parameters
  --datasets                   [string]  The path to datasets that have already been set up
  --species                    [string]  Name of species for species-specific dataset to use

Optional Parameters
  --coverage                   [integer] Reduce samples to a given coverage [default: 100]
  --genome_size                [string]  Expected genome size (bp) for all samples, a value of '0' will disable read error correction and read subsampling,
                                         otherwise estimate with Mash [default: 0]
  --outdir                     [string]  Base directory to write results and Nextflow related outputs to [default: ./]
  --run_name                   [string]  Name of the directory to hold results (e.g. ${params.outdir}/${params.run_name}/<SAMPLE_NAME> [default:
                                         bactopia]

Helpful Parameters
  --available_datasets         [boolean] Print a list of available datasets found based on location given by `--datasets`
  --example_fastqs             [boolean] Print an example FOFN expected from `--fastqs`
  --help_all                   [boolean] An alias for --help --show_hidden_params
  --version                    [boolean] Display version text.

!! Hiding 157 params, use --show_hidden_params (or --help_all) to show them !!
------------------------------------------------------
If you use bactopia for your analysis please cite:

* Bactopia
  https://doi.org/10.1128/mSystems.00190-20

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x

* Software dependencies
  https://github.com/bactopia/bactopia/blob/master/CITATIONS.md
------------------------------------------------------
```



### Bactopia related additions
1. `Utils.groovy` changes
    1. `fileExists()` - return true if file exists else false
    2. `fileNotFound()` - return 1 if file does not exist else 0
    3. `fileNotGzipped()` return 1 if a file is not GZipped else 0
2. `NfcoreSchema.groovy` changes
    1. Added `header` field to print the things like: `### For Procesessing Multiple Samples` 
    2. Added `paramsRequired` to only print the `required_parameters` section
    3. Added `--help_all` to act as an alias for `--help --show_hidden_params`
    3. If parameter has choices (e.g. `--assembler`) print available choices (will follow up with nf-core folks to see if they want a PR)
```
# nf-core output
* --assembler: sksks is not a valid enum value (sksks)

# bactopia output
* --assembler: 'sksks' is not a valid choice (Available choices: skesa, megahit, spades, velvet)
```

### More Examples
```
nextflow run main.nf --coverage not_an_integer
N E X T F L O W  ~  version 21.04.0
Launching `main.nf` [nice_pare] - revision: 555ddf8329

ERROR: Validation of pipeline parameters failed!


* --coverage: expected type: Number, found: String (not_an_integer)


```

```
nextflow run main.nf --SE bactopia
... TRUNCATED ...
------------------------------------------------------
One or more required parameters are missing, please check and try again.


Required Parameters
  ### For Procesessing Multiple Samples
  --fastqs                     [string]  A FOFN with sample names and paths to FASTQ/FASTAs to process

  ### For Processing A Single Sample
  --R1                         [string]  First set of compressed (gzip) paired-end FASTQ reads (requires --R2 and --sample)
  --R2                         [string]  Second set of compressed (gzip) paired-end FASTQ reads (requires --R1 and --sample)
  --SE                         [string]  Compressed (gzip) single-end FASTQ reads  (requires --sample)
  --hybrid                     [boolean] Treat `--SE` as long reads for hybrid assembly.  (requires --R1, --R2, --SE and --sample)
  --sample                     [string]  Sample name to use for the input sequences

  ### For Downloading from SRA/ENA or NCBI Assembly
  **Note: Downloaded assemblies will have error free Illumina reads simulated for processing.**
  --accessions                 [string]  A file containing ENA/SRA Experiment accessions or NCBI Assembly accessions to processed
  --accession                  [string]  Sample name to use for the input sequences

  ### For Processing an Assembly
  **Note: Assemblies will have error free Illumina reads simulated for processing.**
  --assembly                   [string]  A assembled genome in compressed FASTA format. (requires --sample)

!! Hiding 1 params, use --show_hidden_params (or --help_all) to show them !!
------------------------------------------------------
Output directory (.//bactopia) exists, Bactopia will not continue unless '--force' is used.


ERROR: Validation of pipeline parameters failed!
Please correct to continue
```

```
nextflow run main.nf --SE bactopia --sample lll
... TRUNCATED ...
------------------------------------------------------
* --SE: Please verify "bactopia" is compressed using GZIP


Output directory (.//bactopia) exists, Bactopia will not continue unless '--force' is used.


ERROR: Validation of pipeline parameters failed!
Please correct to continue
```

I think at this point I'm going to get a basic workflow without any datasets, then work on cleaning up the Dataset imports and validations

Also, Bactopia has a lot of parameters!  `!! Hiding 157 params, use --show_hidden_params (or --help_all) to show them !!`

**Comment by rpetit3** [comment]
V2 is getting super close!  I've linked the datasets into the DSL2 framework (also fixed an issue with the cache when using `-resume`).

```
[f9/8fb0ff] process > BACTOPIA:GATHER_SAMPLES (SRX4563634)                      [100%] 1 of 1, cached: 1 ✔
[18/57e2ee] process > BACTOPIA:QC_READS (SRX4563634)                            [100%] 1 of 1, cached: 1 ✔
[aa/2f7e80] process > BACTOPIA:ASSEMBLE_GENOME (SRX4563634)                     [100%] 1 of 1, cached: 1 ✔
[45/67a199] process > BACTOPIA:ASSEMBLY_QC (SRX4563634 - quast)                 [100%] 2 of 2, cached: 1 ✔
[dd/e48402] process > BACTOPIA:ANNOTATE_GENOME (SRX4563634)                     [100%] 1 of 1 ✔
[c5/85af3b] process > BACTOPIA:MINMER_SKETCH (SRX4563634)                       [100%] 1 of 1, cached: 1 ✔
[54/c82daa] process > BACTOPIA:ANTIMICROBIAL_RESISTANCE (SRX4563634)            [100%] 1 of 1 ✔
[83/bb0029] process > BACTOPIA:ARIBA_ANALYSIS (SRX4563634 - vfdb_core)          [100%] 2 of 2, cached: 2 ✔
[f9/e25709] process > BACTOPIA:MINMER_QUERY (SRX4563634 - sourmash-genbank-k31) [100%] 4 of 4, cached: 4 ✔
[2f/de1413] process > BACTOPIA:BLAST (SRX4563634 - blastn)                      [100%] 3 of 3 ✔
[27/a09458] process > BACTOPIA:CALL_VARIANTS (SRX4563634 - GCF_000009645)       [100%] 3 of 3, cached: 3 ✔
[99/750675] process > BACTOPIA:MAPPING_QUERY (SRX4563634)                       [100%] 1 of 1, cached: 1 ✔
[34/177ff6] process > BACTOPIA:SEQUENCE_TYPE (SRX4563634 - default)             [100%] 1 of 1 ✔

    Bactopia Execution Summary
    ---------------------------
    Command Line    : nextflow run ../main.nf --accession SRX4563634 --datasets /home/robert_petit/shannon-wgs/datasets/ --species 'Staphylococcus aureus' -profile singularity --genome_size median -resume
    Resumed         : true
    Completed At    : 2021-09-11T05:20:47.912022Z
    Duration        : 4m 10s
    Success         : true
    Exit Code       : 0
    Error Report    : -
    Launch Dir      : /home/robert_petit/bactopia/temp

Completed at: 11-Sep-2021 05:20:48
Duration    : 4m 10s
CPU hours   : 2.0 (71.6% cached)
Succeeded   : 7
Cached      : 15
```

I realize a lot of the code is hidden in the modules, but this is the Bactopia Workflow in its entirety. Its so much cleaner, and much easier to work with now

```
workflow BACTOPIA {
    print_efficiency(RESOURCES.MAX_CPUS) 
    datasets = setup_datasets()

    // Core steps
    GATHER_SAMPLES(create_input_channel(run_type, datasets['genome_size']))
    QC_READS(GATHER_SAMPLES.out.raw_fastq)
    ASSEMBLE_GENOME(QC_READS.out.fastq_assembly)
    ASSEMBLY_QC(ASSEMBLE_GENOME.out.fna, Channel.fromList(['checkm', 'quast']))
    ANNOTATE_GENOME(ASSEMBLE_GENOME.out.fna, Channel.fromPath(datasets['proteins']), Channel.fromPath(datasets['training_set']))
    MINMER_SKETCH(QC_READS.out.fastq)

    // Optional steps that require datasets
    // Species agnostic
    ANTIMICROBIAL_RESISTANCE(ANNOTATE_GENOME.out.annotations, datasets['amr'])
    ARIBA_ANALYSIS(QC_READS.out.fastq, datasets['ariba'])
    MINMER_QUERY(MINMER_SKETCH.out.sketch, datasets['minmer'])

    // Species Specific
    BLAST(ASSEMBLE_GENOME.out.blastdb, datasets['blast'])
    CALL_VARIANTS(QC_READS.out.fastq, datasets['references'])
    MAPPING_QUERY(QC_READS.out.fastq, datasets['mapping'])
    SEQUENCE_TYPE(ASSEMBLE_GENOME.out.fna_fastq, datasets['mlst'])
}
```

Next I will implementment a species-specific workflow (e.g. Staphopia). Then I think we'll be to the point that we can tidy up and start prepping for the V2 release

**Comment by rpetit3** [comment]
Small update - I've incorporated nf-core/modules usage of the `meta` variable to store general info (e.g. id, single_end, etc...). I've always wanted to get away from transporting multiple `val` inputs/outputs between processes and this does that.

I updated the tests for this change, and I've also implemented this in GitHub Actions. So we can now quickly run all the tests that we're making

**Comment by rpetit3** [comment]
`mf-core/modules` is going through some pretty exciting times at the moment related to a new change in version capturing (https://github.com/nf-core/modules/pull/665). These changes are likely to delay some submissions to `nf-core/modules`. 

I'm also finding come cases where the SARS-CoV-2 test data doesn't work well with bacterial tools. To address this I've submitted a PR (https://github.com/nf-core/test-datasets/pull/344) to add some basic bacterial test-data to the `modules` branch on `nf-core/test-datasets`.

**Comment by rpetit3** [comment]
Small update: 11 of 18 bactopia tools have either been merged into `nf-core/modules` or have an open PR. I think that's pretty awesome! Thank you for the help @abhi18av!

I'm hoping to get a few more PRs open tomorrow.

**Comment by rpetit3** [comment]
Thank you for the feedback @Mxrcon!

_You think that automated use of the bactopia tools will be a standard on the V2? This seems to be a good addition as the v2 is moving in the direction of species specific workflows._

I think yes, but we'll have to strike a balance. There are some tools (e.g. `AgrVATE`, `spaTyper`) that are super quick and computationally cheap. For these types of tools I say include them! 

But if certain species-specific tools have significant runtime or resource requirements, then I think those should be executed separately.

**Comment by rpetit3** [comment]
@simone-pignotti  totally agree with @abhi18av 's points

I think for the initial v2 release we might not be quite ready for direct import of bactopia modules (eg. `assemble_genome`) into your own workflows. However I have been thinking about ways to disentangling the modules and configs to make them portable similar to `nf-core/modules`.

**Comment by rpetit3** [comment]
I think we are getting super close, to being ready for v2 release.

I was able to submit a few more PRs to nf-core/modules (`ectyper`, `tbprofiler`, `clonalframeml`, `fastq-scan`, `ncbi-genome-download`).  In stead of waiting on their review, I went ahead and implemented them into Bactopia.

I've converted almost all of the v1 Bactopia Tools to DSL2. It's a very manual process at the moment, but not too tedious and easy to work with (for myself atleast). But I am grouping the Bactopia Tools into two groups:

1. `Subworkflows` - pieces multiple tools together into a single workflow (e.g. pangenome analysis, staphtyper)
2. `Modules` - runs a single tool (e.g. `ectyper`) on multiple samples and merges the results.

There is a parameter `--list_wfs` that will print available workflows. This is also all handled by Nextflow and config files.

```
nextflow run ../main.nf --list_wfs
N E X T F L O W  ~  version 21.04.0
Launching `../main.nf` [infallible_mcclintock] - revision: 49a1306a48


---------------------------------------------
   _                _              _
  | |__   __ _  ___| |_ ___  _ __ (_) __ _
  | '_ \ / _` |/ __| __/ _ \| '_ \| |/ _` |
  | |_) | (_| | (__| || (_) | |_) | | (_| |
  |_.__/ \__,_|\___|\__\___/| .__/|_|\__,_|
                            |_|
  bactopia v1.7.1
---------------------------------------------
Below are a list of workflows you can call using the --wf parameter.

Bactopia
  bactopia (default)       Bactopia is a flexible pipeline for complete analysis of bacterial genomes.
  staphopia                Staphopia is a flexible pipeline for complete analysis of Staphylococcus aureus genomes.

Bactopia Tools
Bactopia Tools can include multiple tools (Subworkflows) or a single tool (Modules).

Subworkflows
  staphtyper               Includes AgrVATE, SpaTyper and Staphpopia SCCmec for Staphylococcus aureus

Modules
  agrvate                  Rapid identification of Staphylococcus aureus agr locus type and agr operon variants
  ectyper                  In-silico prediction of Escherichia coli serotype
  fastani                  fast alignment-free computation of whole-genome Average Nucleotide Identity (ANI)
  hicap                    Identify cap locus serotype and structure in your Haemophilus influenzae assemblies
  kleborate                Screen for MLST, sub-species, and other Klebsiella related genes of interest
  mashtree                 Create a trees using Mash distance
  pirate                   Pangenome toolbox for bacterial genomes
  prokka                   Whole genome annotation of small genomes (bacterial, archeal, viral)
  roary                    Rapid large-scale prokaryote pangenome analysis
  spatyper                 Computational method for finding spa types in Staphylococcus aureus
  staphopiasccmec          Primer based SCCmec typing of Staphylococcus aureus genomes

--------------------------------------------------------------------
If you use bactopia for your analysis please cite:

* Bactopia
  https://doi.org/10.1128/mSystems.00190-20

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x

* Software dependencies
  https://github.com/bactopia/bactopia/blob/master/CITATIONS.md
--------------------------------------------------------------------
```

I'll have to document this process, but I think for power users, it should be pretty straight forward.

I've set in place a path to have the documentation autogenerated from `params.json` files. Which will be super useful for keeping things in sync on the docs side.

**Comment by abhi18av** [comment]
This is awesome - I am learning a lot from the DSl1 -> DSL2 (+ nf-core modules) in Bactopia for some other workflows I've written. Thanks for sharing the updates @rpetit3! 😊

**Comment by rpetit3** [comment]
Some of these aren't quite ready for V2, or need v2 before they can be accomplished. For now I'm going to put them here

### Implement `pytest` for testing
I'd like to create a suite of tests that are operated by `pytest` and `pytest-workflows`. The `nf-core/modules` team has a [framework](https://github.com/nf-core/modules#running-tests-manually) that can be extended to Bactopia.

- Create Tests for Bactopia Tool Modules
    - [ ] `gtdb`
    - [ ] `phyloflash`


### Additional Ways to Run Bactopia
Bactopia, due to Nextflow, can be run on many different environments. But I would like to see it be made easier to run on certain platforms. For example, [Terra](https://terra.bio/) and [CGC](https://www.cancergenomicscloud.org/), but also for advanced cloud users. For these, it might be best to run them as "Nextflow within Nextflow" in order to reduce the number of times instances are started and stopped. This is because we don't want it taking longer to set up an instance that it does to actually run the task.

- Platforms
    - [ ] Terra.bio
    - [ ] Cancer Genomics Cloud
    - [ ] Dockstore 
    - [ ] AWS Cloud Formation
    - [ ] NF-Tower

### Additional Bactopia Tools
There are some Bactopia Tools I would like to get added into v2. Good thing is DSL2 makes it easier to add new tools in the future.

- Tools to Add
    - [ ] ChewBBACA for cgmlst https://github.com/bactopia/bactopia/issues/242
    - [ ] PopPUNK https://github.com/bactopia/bactopia/issues/221
    - [ ] antiSmash https://github.com/bactopia/bactopia/issues/157
    - [x] ~~GET_HOMOLOGUES/GET_PHYLOMARKERS https://github.com/bactopia/bactopia/issues/159#issuecomment-884900884~~ _not available on bioconda_
    - [ ] Phylophlan https://github.com/bactopia/bactopia/issues/248
- Suggestions for Tools
    - [ ] PIRATE/Roary improvements https://github.com/bactopia/bactopia/issues/159

**Comment by rpetit3** [comment]
Happy to report v2 has been released https://github.com/bactopia/bactopia/releases/tag/v2.0.0

Thank you very much @Mxrcon and @abhi18av for your help, and everyone for your feedback! Super excited to see where Bactopia goes from here!
